### PR TITLE
Initial Evaluation Visualization

### PR DESCRIPTION
**Pull request by nicolepeverley**
This is an initial PR to predict a set of output heatmaps for one given data generated image.

Example command to run `evaluation.py` (make sure the model json and model weights are in the current directory):

`python evaluation.py --model-json hpe_hourglass_stacks_04_batchsize_012.json --model-weights hpe_epoch39_val_loss_23851522048.0000_train_loss_0.0192.hdf5 --subset 0.1 --hourglass 4`

I've been writing the heatmap images to local disk and overlayed the 17 outputted heatmaps on top of each other into `result.png`, but they tend to cancel each other out, we might want to find a better way of converting this output back into keypoints before we merge them.

### CODE REVIEW

**Comment by robertklee** [comment]
Looks great @nicolepeverley ! yeah overlaying all the heat maps might be an issue. I think we shouldn't stack the _heatmaps_ together yet. At least, not until our model does something. I think the most useful would be a tiling of outputs in one of the following forms:

1. for each output layer (including intermediate layers), generate a set of all 17 outputs in a row (`np.hstack` should do the trick), followed by the corresponding ground truths on the row below (maybe even overlay the ground truth on the original image).
2. similar to 1, but the first row has the ground truths, and then each layer below is the output from layer 0 to layer n (which is the final output). Basically, 1 but vertically stack each row (`np.vstack` I think).

I think (2) would be better since we can see the progression of predictions across each HG stack. @julianrocha thoughts?

The image names ideally should keep track of the epoch information. Easiest way is to probably find some way to reuse the filename since it should be `..._epoch_{:02d}_...`.

File: evaluation.py
Comment (julianrocha) [inline evaluation.py]: `load_and_filter_annotations` returns `train_df, val_df` so for now we would want to take the second df. Also, we will probably wanna modify `load_and_filter_annotations` to return 3 dfs, `train_df, val_df, test_df` and we would use `test_df` here. I don't think subset argument is necessary here unless we wanna do evaluations on a subset of the test set.

File: evaluation.py
Comment (julianrocha) [inline evaluation.py]: We should figure out why this reshaping is necessary. I would have expected the model to be able to take a single image of shape `(256, 256, 3)`. Is the 1 related the batch size that was used when training the model? If so that seems werid.

File: evaluation.py
Comment (julianrocha) [inline evaluation.py]: Another 1 in the shape that confuses me.

File: evaluation.py
Comment (julianrocha) [inline evaluation.py]: More of quality of life comment, but were you able to get any success with matplot working in Collab so that you can view image overlays in a notebook instead of needing to write to disk?

File: evaluation.py
Comment (robertklee) [reply evaluation.py]: We may just want to `load_and_filter_annotations` to optionally return test, since we don't want to download test2017.zip on the colab VM

File: evaluation.py
Comment (robertklee) [reply evaluation.py]: yeah I think we need the reshape due to batch size. Although I think `numpy.expand_dims` is better than a reshape here.

**Comment by julianrocha** [comment]
Looks great so far. I agree with @robertklee that there is gonna be quite a bit of work to get a nice plot setup that captures the model output in an understandable way. I think in an ideal world option (2) that Robert suggested would give us the most insight. In all cases, overlaying a heatmap over an image rather than plotting the heatmap by itself would be preferable. Obviously, that's gonna take some time to configure perfectly so we can start with just one joint, or the last hourglass layer and work from there. My only suggestion would be to change the plots to be generated online without needing to write to disk as it would be quicker to try different images but this is contingent on us figuring out the matplot versioning issues that are plaguing some of our local python env setups. That being said, some of these 17kp*(8hg + 1 gt) grids of heatmap plots are probably gonna be massive, so an image file that can be zoomed on may be easier to work with. Your call.

File: evaluation.py
Comment (nicolepeverley) [reply evaluation.py]: So I tried testing this with batch size not equal to 1 and it still needs the shape to be (1, 256, 256, 3) for some reason.

File: evaluation.py
Comment (nicolepeverley) [reply evaluation.py]: So I'm currently able to view all plots that are shown in Collab/when saved to Drive but not locally unless saved to disk, I'm not sure if everyone else is having the same issues.

File: evaluation.py
Comment (julianrocha) [reply evaluation.py]: @nicolepeverley have you tried on different models that have been trained with different batch sizes? In all cases is it always 1?

Comment (robertklee) [reply evaluation.py]: It's because the keras model can process multiple images at once [Source](https://stackoverflow.com/questions/50723287/meaning-of-batch-size-in-model-evaluate):
> The evaluate function of Model has a batch size just in order to speed-up evaluation, as the network can process multiple samples at a time, and with a GPU this makes evaluation much faster. I think the only way to reduce the effect of this would be to set batch_size to one.

File: evaluation.py
Comment (julianrocha) [reply evaluation.py]: If it ends up being easier, you can always try moving the visualization logic to a notebook. That's what I did for `coco-explorer` And just work in google collab

File: evaluation.py
Comment (julianrocha) [reply evaluation.py]: Oh perfect that explains it then ty

Comment (robertklee) [reply evaluation.py]: (I didn't know this until now haha)

File: evaluation.py
Comment (julianrocha) [reply evaluation.py]: Nvm makes sense as per Roberts batch evaluation comment

File: evaluation.py
Comment (robertklee) [reply evaluation.py]: If you've used the `# %%` short form for a normal `.py` file, you can right click on the file in VS code and convert to a python notebook too

**Comment by nicolepeverley** [comment]
Stacked output of each hourglass set of heatmaps and ground truth heatmaps for a given image:

evaluation

**Comment by nicolepeverley** [comment]
The last commit stacks the output for better visualization, I'm going to check to see if these images need to be saved to disk to accomplish this or not 👍 @julianrocha

**Comment by wanze-zhang** [comment]
Looks great!!!

**Comment by nicolepeverley** [comment]
@Airborne95 and I discussed evaluation moving forward and we're going to update the architecture to look similar to data_generator.py and data_generator_tests.py stay tuned!

**Review by Airborne95 — CHANGES_REQUESTED** [review]
Per my discussion with @nicolepeverley, this file will be refactored into a class similar to `data_generator.py`. This pattern will make it easier to run evaluations from any type of notebook.

File: evaluation.py
Comment (Airborne95) [inline evaluation.py]: **Nitpick**
@nicolepeverley I think you mentioned this is a private method so you could change the name to `__hstack_images`.

File: evaluation.py
Comment (Airborne95) [inline evaluation.py]: **Nitpick**
@nicolepeverley I think you mentioned this is a private method so you could change the name to `__vstack_images`.

**Review by Airborne95 — APPROVED** [review]
Its looking really good! I think you did a great job refactoring it into a class 🚀

File: evaluation.py
Comment (robertklee) [reply evaluation.py]: Good call. I'd avoid `__` prefixes though because double underscores have a special meaning in python. Convention is single underscore `_` to hint private methods/variables

Comment (robertklee) [reply evaluation.py]: > The use of double underscore (__) in front of a name (specifically a method name) is not a convention; it has a specific meaning to the interpreter.

> Python mangles these names and it is used to avoid name clashes with names defined by subclasses.

> This is also called name mangling — the interpreter changes the name of the variable in a way that makes it harder to create collisions when the class is extended later.

File: evaluation.py
Comment (robertklee) [inline evaluation.py]: nit: calling this `model` is a bit confusing. Is it the `model_arch` or the model class itself (aka an instance of Hourglass class)?

File: run_evaluation.py
Comment (robertklee) [inline run_evaluation.py]: Is it possible to generate the fully stacked ones without requiring the other ones to be saved to disk? A few reasons, 1) mainly so it's faster to run as everything is kept in memory, and 2) I read that Google Colab has resource limits on how much you can write to and from Google Drive. They penalize small file read/writes more than big ones.

File: run_evaluation.py
Comment (robertklee) [inline run_evaluation.py:42]: This is great! It looks like once we have the single image pipeline working, it's a matter of calling 
```
for X_batch, y_stacked in generator:
```

File: evaluation.py
Comment (robertklee) [inline evaluation.py:8]: I'm loving this class refactor!

**Review by robertklee — APPROVED** [review]

File: run_evaluation.py
Comment (robertklee) [reply run_evaluation.py]: I doubt that this would trigger the resource limits, but if we're running a script that generates evaluation it for a ton of epochs and a ton of different model training sessions, we might want to play it safe. Also, if we end up calling this in the `on_epoch_end` callback, we don't want to be crashing in the middle of training.

Comment (nicolepeverley) [reply run_evaluation.py]: I'll check on this! I tried using the numpy arrays before and ran into some issues but I'll double check cause it would be much faster 👍

Comment (robertklee) [reply run_evaluation.py]: if you get all of the hg predictions in a list or np array, you can use `np.hstack([list here])` :)

Comment (nicolepeverley) [reply run_evaluation.py]: Yeah np.hstack and np.vstack work great for the output heatmaps, but the ground truth heatmaps are using https://github.com/LinShanify/HeatMap so I'm trying to see if I can convert them to np arrays!

Comment (robertklee) [reply run_evaluation.py]: Sweet! There might be a method in one of the image processing libraries that act as a file buffer to write pictures to. Worst case, since they don't change between layers, we could generate it once and use everywhere too.

File: evaluation.py
Comment (julianrocha) [inline evaluation.py:38]: Succinct methods are a beauty to read nicely done.

File: run_evaluation.py
Comment (julianrocha) [reply run_evaluation.py]: I agree with @robertklee  about only saving the final fully stacked image to disk for simplicity and efficiency. @nicolepeverley if the HeatMap submodule is feeling restrictive, keep in mind we have the option to customize it, so we could add a flag to the plot method that just returns the raw NumPy array instead of trying to plot it (just one idea).

Comment (julianrocha) [reply run_evaluation.py]: It looks like you were considering this already

File: evaluation.py
Comment (julianrocha) [inline evaluation.py]: Curious what the final use case is for this evaluator. Will we always be trying to plot one and only one image? If we wanna one day plot many, then filenames will need some kind of ID for what image they belong to (annotation ID maybe?). Might be fine if the other files get overwritten since they are intermediate results, but this final result will need an ID. That being said if we only ever intend to use this on one image at a time then this is fine.

**Review by julianrocha — APPROVED** [review]
Most of my comments were related to how files are handled. Otherwise, this is great, gonna give us huge insight into how the model handles specific images and how the heatmaps change across hourglasses (<- useful for picking the best hg number). Also a huge shoutout to the refactoring that happened here, the difference in readability was night and day compared to the first time I read this PR so great stuff.

File: evaluation.py
Comment (robertklee) [inline evaluation.py]: small nit: 
```suggestion
        X = np.expand_dims(X, axis=0) # add "batch" dimension of 1 because predict needs shape (1, 256, 256, 3)
```

File: evaluation.py
Comment (robertklee) [inline evaluation.py:81]: It looks like we don't need to pass in files anymore, right?

File: run_evaluation.py
Comment (robertklee) [inline run_evaluation.py]: Do we need to save still?

File: run_evaluation.py
Comment (robertklee) [inline run_evaluation.py]: would it be better for `eval` to have one function to call that calls these methods?

**Review by robertklee — APPROVED** [review]
Overall looks great!

**Comment by nicolepeverley** [comment]
Okay so the reason I was having trouble vertically stacking/concatenating the predicted heatmaps and ground truth heatmaps  is because the ground truth heatmaps get returned as 3 dimensional np arrays (our predicted heatmaps are 2 dimensional np arrays). I'm not sure how to get around this so for now I've left it that if you want to see the predicted heatmaps and ground truth heatmaps stacked together you'll have to save them!

File: notebooks/input_pipeline_demo.ipynb
Comment (nicolepeverley) [inline notebooks/input_pipeline_demo.ipynb:1025]: This is the only change here for this file!

**Review by robertklee — APPROVED** [review]
LGTM! This isn't a blocker but one possible option to temporarily save an image in-memory is [this StackOverflow answer](https://stackoverflow.com/questions/33916409/how-to-save-image-in-memory-and-upload-using-pil)

```
temp = BytesIO()
im.save(temp, format="png")
```

**Comment by robertklee** [comment]
We can explore this tiny QOL improvement after merging though

File: data_generator_tests.py
Comment (julianrocha) [inline data_generator_tests.py:79]: ty for fixing the dependencies

**Review by julianrocha — APPROVED** [review]
Juicy

**Pull request merged**
